[CD] Annotate linux/arm64 cuda wheels with consistent nvidia dependencies

tmm1 opened the pull request:
This resolves issues installing torch nightly wheels into a `uv sync`-generated `.venv`

The root cause is that the x64 and arm64 cuda nightly wheels have inconsistent metadata. This can be seen comparing `generated-linux-aarch64-binary-manywheel-nightly.yml` and `generated-linux-binary-manywheel-nightly.yml`

`uv` expects consistency:

https://github.com/astral-sh/uv/issues/10693
>Frankly, it's really not ideal that they change their dependencies from wheel to wheel.
>They could still put the dependencies there with the same platform markers they're using in the other wheel though... 🤷‍♀

https://github.com/astral-sh/uv/issues/10119#issuecomment-2559898792
>I think this is something that basically has to be solved by PyTorch. The issue is that the wheels for `2.6.0.dev20241222+cu126` don't have consistent metadata, and it's a fundamental assumption of uv that the metadata for a given version _is_ consistent.

To resolve this, I modified the arm64 nightly build workflow to add two new `PYTORCH_EXTRA_INSTALL_REQUIREMENTS` entries, under `manywheel-py3_11-cuda-aarch64-build` and `manywheel-py3_12-cuda-aarch64-build`. These are based on their equivalents in the x64 workflow for the corresponding python versions.

I used the cuda 12.6 dependencies versions for the nvidia packages, to match the `DOCKER_IMAGE: pytorch/manylinuxaarch64-builder:cuda12.6-main` being used by these jobs.

(The arm64 workflow file already had several `PYTORCH_EXTRA_INSTALL_REQUIREMENTS` entries, under various cpu wheels. I'm not sure why these are there, but I left them as-is.)

tmm1 committed:
[CD] Annotate linux/arm64 cuda wheels with consistent nvidia dependencies

This resolves issues installing torch nightly wheels into a uv venv (astral-sh/uv#10693)

Skylion007 commented:
These are generated files so this chang is a no_op, please modify the actual python JINJA files which edit the yaml

tmm1 commented:
@Skylion007 Thanks for the pointer!

I updated the script to:

- remove the nvidia deps on the aarch64 cpu wheels
- add nvidia deps on the aarch64 cuda wheels

and updated the PR and generated yml files accordingly.

tmm1 committed:
make it more future proof

seemethere commented:
We might need to rebase / regenerate the workflows here

seemethere commented:
@pytorchbot merge

seemethere committed:
fix lint

Signed-off-by: Eli Uriegas <eliuriegas@meta.com>

seemethere commented:
/easycla

atalman commented:
This is attempt to fix https://github.com/pytorch/pytorch/issues/146679 However we need to have CUDA dependencies as part of cpu aarch64 binaries as well

lukaszgryglicki commented:
/easycla

lukaszgryglicki commented:
/easycla

tmm1 committed:
rerun .github/scripts/generate_ci_workflows.py

tmm1 committed:
leave cpu wheels as-is

reverts "remove the nvidia deps on the aarch64 cpu wheels"

Co-authored-by: Andrey Talman <atalman@fb.com>

tmm1 commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -429,7 +430,7 @@ def generate_wheels_matrix(
                             if gpu_arch_type == "xpu"
                             else PYTORCH_EXTRA_INSTALL_REQUIREMENTS["12.4"]
-                            if os != "linux"
+                            if os != "linux" and arch_version != "cpu-aarch64"
```
Okay. Does it make sense to use the 12.4 versions there?

tmm1 committed:
rerun .github/scripts/generate_ci_workflows.py

atalman commented:
@tmm1 Looks good, let us build this wheel and validate the METADATA section

atalman commented:
Looks good:
```
Requires-Dist: nvidia-cuda-nvrtc-cu12==12.6.77; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cuda-runtime-cu12==12.6.77; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cuda-cupti-cu12==12.6.80; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cudnn-cu12==9.5.1.17; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cublas-cu12==12.6.4.1; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cufft-cu12==11.3.0.4; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-curand-cu12==10.3.7.77; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cusolver-cu12==11.7.1.2; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cusparse-cu12==12.5.4.2; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cusparselt-cu12==0.6.3; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-nccl-cu12==2.25.1; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-nvtx-cu12==12.6.77; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-nvjitlink-cu12==12.6.85; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cufile-cu12==1.11.1.6; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: pytorch-triton==3.2.0+git4b3bb1f8; platform_system == "Linux" and platform_machine == "x86_64"
```

atalman reviewed (approved):
lgtm

atalman commented:
@pytorchmergebot merge -f "builds and lint looks good"

jjh42 commented:
I'm still struggling to get uv to install torch nightly (uv pip works but I mean from pyproject.toml with sync). Does anyone have an example now that this is done?
